### DISCUSSION TITLE
Creature smooth rotation

### DISCUSSION THREAD

**Discussion by Yehonal** [azerothcore/azerothcore-wotlk#4401] [General]
Since I'm not 100% sure it's an issue or not, I'm creating a topic here to collect info and discussions about the creature smooth rotations on waypoints and random movement.

## The retail behaviour

On retail (both classic and shadowlands) the creatures turn smoothly when they change direction while they are following waypoints or they are in idle random movement.

https://www.youtube.com/watch?v=zDbK-HHO7gk

## The behaviour in retail during the 2009 (wotlk)

Some people, however, noticed that in a video  from the 2009 the creatures were following a non smooth path: https://www.youtube.com/watch?v=O_5gRE8vy0k&list=PLF5A2ABB39D69E22C&t=29s

## AC behaviour

The AC behaviour right now is basically the same of the 2009 video, however, we have not enough resources yet to understand which was the correct behaviour. 

## Other info and considerations

Even though the official classic expansion should resemble the version they had back in times, it's also possible that they upgraded some parts of the client and the server with some improvements from the latest expansions.

In any case, the only possible way to implement it correctly would be to find the correct information to send to the client to properly show the "turning" animation. There are some partially working experiments done but they are using workarounds that cannot be considered a solution.

**References:**

https://github.com/cmangos/issues/issues/2050

https://github.com/azerothcore/azerothcore-wotlk/pull/2796

https://github.com/TrinityCore/TrinityCore/issues/22863

**Comment by Yehonal** [comment]
To test a global smooth for all the creatures of the world, you can change this line in ObjectMgr.cpp

![image](https://user-images.githubusercontent.com/147092/106445855-d6a9ab80-647f-11eb-8a45-c4febfbb88ed.png)

**NOTE**: you MUST clean the client cache

**Reply by virgo77** [reply]
Hi!
I was thinking about asking about it, so thanks for that :)
I will try your change this week. But I have to know if this modification can be "erased" by a Core Git Pull Update? If yes, do you think that it is possible to make the change by default? So we could have a better and smooth NPC movement for time :)
I will reply here if it's working fine, or not
Have a good day, all!
V.

**Reply by Yehonal** [reply]
This can't be a definitive change because it's just a hack to experiment it but it doesn't work properly

**Reply by alchem1ster** [reply]
> So we could have a better and smooth NPC movement

We need to implement some kind of interpolation for real smooth movement instead

**Comment by alchem1ster** [comment]
I found this: https://youtu.be/74_cvkTGriE?t=420
So, in the past, NPCs didn't move smoothly :)

**Reply by Yehonal** [reply]
Thank you! That's definitely interesting.

It also means that the current classic version of the game on retail has been improved somehow.

Answer approved ✅

**Reply by alchem1ster** [reply]
> has been improved somehow

There are many variants :)
1. Spline interpolation;
2. More waypoints  to smoother movements;
3. Changing the animation on the client side.

**Reply by Kitzunu** [reply]
Well they didnt really have the intention of classic to be as back in the day @Yehonal as they have fixed many "retail bugs". As well as the client being based on legion if im not misstaken
